eZ/Publish/Core/Persistence/Cache/Tags/TagGenerator.php

```diff
@@ -0,0 +1,28 @@
+<?php
+
+/**
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace eZ\Publish\Core\Persistence\Cache\Tags;
+
+class TagGenerator implements TagGeneratorInterface
+{
+    private const PLACEHOLDER = '-%s';
+    private const SUFFIX_SEPARATOR = '-';
+
+    public function generate(string $patternName, array $values = [], bool $isPrefix = false): string
+    {
+        $pattern = constant(TagIdentifierPatterns::class . '::' . strtoupper($patternName));
```

Comment:
I don't like this magic here. It's difficult to maintain and error-prone. What I would do is place it in lazy-loaded Yaml configuration - e.g. using static method to store already loaded result. Configuration file path should be placed in a container parameter so it's "overridable".

But before any action is taken - POV ping @ezsystems/php-dev-team

Reply:
> I don't like this magic here. It's difficult to maintain and error-prone.
I agree.  

> What I would do is place it in lazy-loaded Yaml configuration - e.g. using static method to store already loaded result. Configuration file path should be placed in a container parameter so it's "overridable".

Partially agree. Loading and parsing YAML on every single request will kill performance. KISS: Store tag patterns as regular DIC parameter and simply injected into Tag Generator.
 
```php
<?php

namespace eZ\Publish\Core\Persistence\Cache\Tags;

use eZ\Publish\Core\Base\Exceptions\InvalidArgumentException;

/**
 * @internal
 */
final class TagGeneratorV2 implements TagGeneratorInterface
{
    /** @var string */
    private $prefix;
    
    /** @var array<string,string> */
    private $patterns;

    public function __construct(string $prefix, array $patterns)
    {
        $this->prefix = $prefix;
        $this->patterns = $patterns;
    }

    public function generate(string $patternName, array $values = [], bool $withPrefix = false): string
    {
        if (isset($this->patterns[$patternName])) {
            $tag = sprintf($this->patterns[$patternName], ...$values);
            if ($withPrefix) {
                $tag = $this->prefix . $tag;
            }

            return $tag;
        }

        throw new InvalidArgumentException('$patternName', 'Undefined tag pattern');
    }
}
```

```yaml
parameters:
    ibexa.persistance.cache.tag_prefix: 'ez-'
    ibexa.persistance.cache.tag_patterns:
        bookmark: 'bg-%s'
        content: 'c-%s'
        content_info: 'ci-%s'
        # ...
        
services:
    # ...
    eZ\Publish\Core\Persistence\Cache\Tags\TagGeneratorV2:
        arguments:
            $prefix: '%ibexa.persistance.cache.tag_prefix%'
            $patterns: '%ibexa.persistance.cache.tag_patterns%'
```

Additionally we should define valid pattern names as const. in TagGeneratorInterface or TagIdentifierPatterns class to be able to easily access them in PHP code (and provide autocompletion in PHPStorm). 

With ^ approach providing BC layer will be just only matter of overriding `ibexa.persistence.cache.tag patterns` parameter value.

Reply:
> Partially agree. Loading and parsing YAML on every single request will kill performance. KISS: Store tag patterns as regular DIC parameter and simply injected into Tag Generator.

+1, idk why I tried to reinvent the wheel, patterns itself should be parameters.

Reply:
What I tried to achieve in the first place was to omit the performance hit, but @adamwojs approach seems to be the best one.